zerver/actions/message_flags.py

```diff
@@ -219,10 +219,6 @@ def do_clear_mobile_push_notifications_for_ids(
     if len(message_ids) == 0:
         return
 
-    # This function supports clearing notifications for several users
-    # only for the message-edit use case where we'll have a single message_id.
-    assert len(user_profile_ids) == 1 or len(message_ids) == 1
-
     messages_by_user = defaultdict(list)
     notifications_to_update = (
         UserMessage.objects.filter(
```

Comment:
```diff
-   # This function supports clearing notifications for several users
-   # only for the message-edit use case where we'll have a single message_id.
-   assert len(user_profile_ids) == 1 or len(message_ids) == 1

    messages_by_user = defaultdict(list)
    notifications_to_update = (
        UserMessage.objects.filter(
            message_id__in=message_ids,
            user_profile_id__in=user_profile_ids,
        )
        .extra(
            where=[UserMessage.where_active_push_notification()],
        )
        .values_list("user_profile_id", "message_id")
    )
```
This query would be expensive when we perform a topic deletion. But that's OK as "topic deletion" is not a frequent operation ?

Reply:
I think we don't really have the right database indexes to make this fast given the quadratic work -- and I'm worried about the request timing out and failing when deleting a topic. Start a thread in #backend about how to do this query; one theory is that we use the `deferred_work` queue for this. A good structure might be:

> If there are both multiple messages and multiple users, enqueue a `deferred_work` job for each user with the set of messages (that structure seems good given that's how the events are organized). 

But even that might be bad (floods the queue) if there's 20k users and 1k messages. The actual index on the usermessage table is really sparse; it's possible that just walking the full index for all usermessages in the organiation is fast enough and this operation will complete, but I'm a bit skeptical, especially given the query plan I expect this to generate which might be quadratic.

This might be a thing that we can't happily do until we add a `UserMessage.realm` column (which we need for sharding) and do a `(realm, user_id, message_id)` partial index.

Reply:
[Started a discussion on CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/clear.20push.20notifications/near/1645830)